core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/request/resource/BaseMetadataResourceConverter.java

```diff
@@ -13,127 +13,125 @@
+import com.b2international.snowowl.core.version.Version;
 import com.b2international.snowowl.core.version.Versions;
+import com.google.common.collect.HashMultimap;
+import com.google.common.collect.Multimap;
+import com.google.common.collect.TreeMultimap;
 
 /**
- * @since 8.0
+ * @since 8.12
  */
-public final class ResourceConverter extends BaseResourceConverter<ResourceDocument, Resource, Resources> {
+public abstract class BaseMetadataResourceConverter<R extends Resource, CR extends PageableCollectionResource<R>> extends BaseResourceConverter<ResourceDocument, R, CR> {
 
 	public static final String ROOT_LABEL = "Root";
 	
 	private static final String MISSING_BUNDLE_TEMPLATE = "<bundle %s>";
 	
 	private final ResourceTypeConverter.Registry converters;
-
-	public ResourceConverter(RepositoryContext context, Options expand, List<ExtendedLocale> locales) {
+	
+	public BaseMetadataResourceConverter(RepositoryContext context, Options expand, List<ExtendedLocale> locales) {
 		super(context, expand, locales);
 		this.converters = context().service(ResourceTypeConverter.Registry.class);
 	}
 	
 	@Override
-	protected RepositoryContext context() {
+	protected final RepositoryContext context() {
 		return (RepositoryContext) super.context();
 	}
-
-	@Override
-	protected Resources createCollectionResource(List<Resource> results, String searchAfter, int limit, int total) {
-		return new Resources(results, searchAfter, limit, total);
-	}
-
+	
 	@Override
-	protected Resource toResource(ResourceDocument doc) {
-		return converters.toResource(doc);
+	protected final R toResource(ResourceDocument doc) {
+		return (R) converters.toResource(doc);
 	}
 	
 	@Override
-	public void expand(List<Resource> results) {
-		if (results.isEmpty() || expand().isEmpty()) {
-			return;
-		}
-
+	public void expand(List<R> results) {
 		expandCommits(results);
 		expandUpdateAtCommit(results);
 		expandResourcePathLabels(results);
 		expandVersions(results, expand(), options -> getLimit(options), locales(), context());
 		
+		expandDependencyUpgrades(results);
+		
 		// expand additional fields via pluggable converters
 		converters.expand(context(), expand(), locales(), results);
 	}
-
-	private void expandUpdateAtCommit(List<Resource> results) {
-		if (expand().containsKey(Resource.Expand.UPDATED_AT_COMMIT)) {
-			// expand updatedAtCommit object for each resource one-by-one
-			results.stream()
-				.forEach(res -> {
-					RepositoryRequests.commitInfos().prepareSearchCommitInfo()
-						.one()
-						.filterByBranch(Branch.MAIN_PATH) // all resource commits go to the main branch
-						.filterByAffectedComponent(res.getId())
-						.setFields(CommitInfo.Fields.DEFAULT_FIELD_SELECTION)
-						.sortBy("timestamp:desc")
-						.build()
-						.execute(context())
-						.first()
-						.ifPresent(res::setUpdatedAtCommit);
-				});
+	
+	protected final void expandDependencyUpgrades(List<R> results) {
+		if (!expand().containsKey("dependencies_upgrades")) {
```

Comment:
It would also be nice to save this in a static field somewhere, like:
https://github.com/b2ihealthcare/snow-owl/blob/4437ec2b51e6850a1092ed1fec5650b529b9bfa4/snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/core/domain/SnomedConcept.java#L113

Reply:
This is a temporary expand key, we need dot notation support here, like this `dependencies.upgrades()` but unfortunately the current expand parser does not like dots in property names and fails to parse the object. A minor refactor is required to make it work so we had to use underscore instead of dot for the separator.